components/calendar/src/ixdtf.rs

```diff
@@ -0,0 +1,310 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::str::FromStr;
+
+use crate::{AnyCalendar, Date, DateTime, Iso, RangeError, Time};
+use ixdtf::parsers::records::{DateRecord, IxdtfParseRecord, TimeRecord};
+use ixdtf::parsers::IxdtfParser;
+use ixdtf::ParserError;
+
+/// An error returned from parsing an IXDTF string to an `icu_calendar` type.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum FromIxdtfError {
+    /// Syntax error in the IXDTF string.
+    Ixdtf(ParserError),
+    /// Value is out of range of the `icu_calendar` type.
+    Range(RangeError),
+    /// The IXDTF is missing fields required for parsing into the chosen type.
+    Missing,
+    /// The IXDTF specifies a calendar unknown to `icu_calendar`.
+    UnknownCalendar,
+}
+
+impl From<RangeError> for FromIxdtfError {
+    fn from(value: RangeError) -> Self {
+        Self::Range(value)
+    }
+}
+
+impl From<ParserError> for FromIxdtfError {
+    fn from(value: ParserError) -> Self {
+        Self::Ixdtf(value)
+    }
+}
+
+impl TryFrom<DateRecord> for Date<Iso> {
+    type Error = RangeError;
+    fn try_from(value: DateRecord) -> Result<Self, Self::Error> {
+        Self::try_new_iso_date(value.year, value.month, value.day)
+    }
+}
+
+impl TryFrom<TimeRecord> for Time {
+    type Error = RangeError;
+    fn try_from(value: TimeRecord) -> Result<Self, Self::Error> {
+        Self::try_new(value.hour, value.minute, value.second, value.nanosecond)
+    }
+}
+
+impl AnyCalendar {
+    #[cfg(feature = "compiled_data")]
+    fn try_from_ixdtf_record(ixdtf_record: &IxdtfParseRecord) -> Result<Self, FromIxdtfError> {
+        let calendar_id = ixdtf_record.calendar.unwrap_or(b"iso");
+        let calendar_kind = crate::AnyCalendarKind::get_for_bcp47_bytes(calendar_id)
+            .ok_or(FromIxdtfError::UnknownCalendar)?;
+        let calendar = AnyCalendar::new(calendar_kind);
+        Ok(calendar)
+    }
+}
```

Comment:
hmm the records feel like ixdtf parser implementation types, so I'd rather not expose conversions from them. `Date` and `Time` are trivial enough to be done manually, and for `AnyCalendar` I don't really see the use case

Reply:
I figured since `DateRecord` and `TimeRecord` are public types, we should have public conversions. But, actually, it puts `ixdtf` into `icu_datetime`'s API, so we should probably at least wait until `ixdtf` has a 1.0 release.